web/postcss/roots-merger.ts

```diff
@@ -0,0 +1,50 @@
+// PostCSS plugin for ../script/clean-css.ts to merge all rules
+// with :root selector into the single rule in the beginning of the file
+
+import { rule } from 'postcss'
+import type { ChildNode, Plugin, Rule } from 'postcss'
+
+export const rootsMerger: Plugin = {
+  postcssPlugin: 'roots-merger',
+  prepare() {
+    let rootNodes = new Map<string, ChildNode>()
+    let rulesToRemove: Rule[] = []
+
+    return {
+      OnceExit(root) {
+        rulesToRemove.forEach(cssRule => cssRule.remove())
+
+        let rootRule = rule({ selector: ':root' })
+
+        rootRule.append(...rootNodes.values())
+        if (rootRule.nodes.length > 0) {
+          root.prepend(rootRule)
+        }
+      },
+      Rule(cssRule) {
+        if (cssRule.selector !== ':root') {
+          return
+        }
+
+        if (cssRule.parent && cssRule.parent.type === 'atrule') {
+          return
+        }
+
+        cssRule.nodes.forEach(node => {
+          if (node.type !== 'decl') {
+            return
+          }
+
+          // remove rule from the map to preserve the rules order
+          if (rootNodes.has(node.prop)) {
+            rootNodes.delete(node.prop)
+          }
+
+          rootNodes.set(node.prop, node)
+        })
+
+        rulesToRemove.push(cssRule)
```

Comment:
Why not remove it here?

Reply:
That's because we run both plugins at a time:
`const cssCleaner = postcss([rootsMerger, varsCleaner])`

If we remove the rule here, `varsCleaner` will not "see" it. This could cause the situation where unused rules will appear in the resulting CSS artifact even after processing by the `varsCleaner`.

We can run plugins one by one, but it seems that it would be more optimal to run them in the current way

Reply:
Got it 🤔